ipaserver/install/ipa_idrange_fix.py

```diff
@@ -0,0 +1,1073 @@
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import logging
+import ldap
+
+from ipalib import api, errors
+from ipapython.admintool import AdminTool
+from ipapython.dn import DN
+from ipapython import ipautil
+
+logger = logging.getLogger(__name__)
+
+
+class IDRange:
+    """Class for ID range entity"""
+
+    def __init__(self, **kwargs):
+        self.last_id: int = None
+        self.last_base_rid: int = None
+        self.last_secondary_rid: int = None
+        self.name: str = kwargs.get('name')
+        self.size: int = kwargs.get('size')
+        self.first_id: int = kwargs.get('first_id')
+        self.base_rid: int = kwargs.get('base_rid')
+        self.secondary_base_rid: int = kwargs.get('secondary_base_rid')
+        self.type: str = kwargs.get('type')
+        self.suffix: str = kwargs.get('suffix')
+        self.dn: str = kwargs.get('dn')
+        if self.first_id is not None and self.size is not None:
+            self.count()
+
+    def count(self) -> None:
+        """Function to calculate last IDs for the range"""
+        self.last_id = self.first_id + self.size - 1
+        if self.type == "ipa-local":
+            self.last_base_rid = (
+                self.base_rid + self.size
+                if self.base_rid is not None
+                else None
+            )
+            self.last_secondary_rid = (
+                self.secondary_base_rid + self.size
+                if self.secondary_base_rid is not None
+                else None
+            )
+
+    def __repr__(self):
+        return f"IDRange(name='{self.name}', \
+type={self.type}, \
+size={self.size}, \
+first_id={self.first_id}, \
+base_rid={self.base_rid}, \
+secondary_base_rid={self.secondary_base_rid})"
+
+    def __eq__(self, other):
+        return self.first_id == other.first_id
+
+
+class IDentity:
+    """Class for ID entity"""
+
+    def __init__(self, **kwargs):
+        self.dn: str = kwargs.get('dn')
+        self.name: str = kwargs.get('name')
+        self.user: str = kwargs.get('user')
+        self.number: int = kwargs.get('number')
+
+    def __str__(self):
+        if self.user:
+            return (f"user '{self.name}', uid={self.number}")
+        return (f"group '{self.name}', gid={self.number}")
+
+    def debug(self):
+        if self.user:
+            return (
+                f"user(username='{self.name}', "
+                f"uid={self.number}, "
+                f"{self.dn})"
+            )
+        return (
+            f"group(groupname='{self.name}', "
+            f"gid={self.number}, "
+            f"{self.dn})"
+        )
+
+    def __eq__(self, other):
+        return self.number == other.number
+
+
+class IPAIDRangeFix(AdminTool):
+    """Tool to analyze and fix IPA ID ranges"""
+
+    command_name = "ipa-idrange-fix"
+    log_file_name = "/var/log/ipa-idrange-fix.log"
```

Comment:
Can you open a healthcheck ticket for this new log? We'll want to verify ownership and permissions of it eventually.

Reply:
opened https://github.com/freeipa/freeipa-healthcheck/issues/335